src/keys/slip10.rs

```diff
@@ -588,3 +664,162 @@ impl Segment for NonHardened {
+    ///     return (child_0, .., child_{address_count - 1})
+    ///
+    /// Star (*) denotes hardening rule specific for key type `K`.
+    ///
+    /// Address space should not overflow, if `k` is the first index such that `address_index + k` overflows (31-bit),
+    /// then only the first `k` children are returned.
+    pub fn derive_address_range<K, S>(
+        &self,
+        m: &Slip10<K>,
+        address_count: usize,
+    ) -> impl ExactSizeIterator<Item = Slip10<K>>
+    where
+        K: hazmat::Derivable + hazmat::WithSegment<S> + hazmat::ToChain<Bip44, Chain = [S; 5]>,
+        S: Segment + TryFrom<u32>,
+        <S as TryFrom<u32>>::Error: core::fmt::Debug,
+    {
+        let chain: [_; 5] = self.to_chain::<K>();
+
+        // maximum number segments is 2^31, trim usize value to fit u32
+        let address_count = core::cmp::min(1 << 31, address_count) as u32;
+
+        // BIP44 conversion rules are strict, the last element is address_index
+        let address_start = chain[4];
+        let hardening_bit: u32 = address_start.into() & HARDEN_MASK;
+        // strip hardening bit as it may interfere and overflow
+        let unhardened_start: u32 = address_start.unharden().into();
+        // this is guaranteed to not overflow and be <= 2^31
+        let unhardened_end: u32 = core::cmp::min(1_u32 << 31, unhardened_start + address_count);
+
+        // this is the final range guaranteed to not overflow address_index space
+        let child_segments = (unhardened_start..unhardened_end).map(move |unhardened_address_index| -> S {
+            let address_index = hardening_bit | unhardened_address_index;
+            // SAFETY: address_index is guaranteed to have the correct hardening as the target type `S`, so unwrap()
+            // can't fail
+            address_index.try_into().unwrap()
+        });
+
+        let mk = if child_segments.len() > 0 {
+            m.derive(chain[..4].iter().copied())
+        } else {
+            // no need to derive mk if there's no child_segments, just use empty/zero one
+            Slip10::new()
+        };
+        mk.children(child_segments)
+    }
+}
+
+impl From<[u32; 5]> for Bip44 {
+    fn from(segments: [u32; 5]) -> Self {
+        let [purpose, coin_type, account, change, address_index] = segments;
+        Self {
+            purpose,
+            coin_type,
+            account,
+            change,
+            address_index,
+        }
+    }
+}
+
+impl From<&Bip44> for [u32; 5] {
+    fn from(bip44_chain: &Bip44) -> [u32; 5] {
+        [
+            bip44_chain.purpose,
+            bip44_chain.coin_type,
+            bip44_chain.account,
+            bip44_chain.change,
+            bip44_chain.address_index,
+        ]
+    }
+}
+
+pub struct Bip44Builder(Bip44);
+
+impl Bip44Builder {
+    pub fn new() -> Self {
+        Self(Bip44::from([0, 0, 0, 0, 0]))
+    }
+    pub fn purpose(mut self, s: u32) -> Self {
+        self.0.purpose = s;
+        self
+    }
```

Comment:
Shouldn't the purpose be hardcoded to 44?
https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#purpose

```suggestion
        Self(Bip44::from([44, 0, 0, 0, 0]))
    }
```

Reply:
Should Bip44::purpose also be private? Should impl From<[u32; 5]> for Bip44 assert that purpose == 4? Or should it be TryFrom?

Reply:
If it can fail it should be try_from
Or another idea, maybe just rename it to Bip32? Because I think that's what it is now

Reply:
> Should Bip44::purpose also be private?

If we always only want to use it for Bip44 I would say yes, no point in having the option to set something else then

Reply:
Removed it. Now it's fixed constant.